### ISSUE TITLE
Support i18n

### ISSUE THREAD

**Issue by tobegit3hub** [mkdocs/mkdocs#211]
_&lt;edit&gt;_

### Note from project maintainers:

There are multiple aspects to supporting multiple languages in MkDocs, each of which is addressed in its own separate issue as listed below. This is the primary issue which ties all of the various aspects together. It is best to discuss each aspect in its related issue. This issue will be closed when all of the related issues are resolved (accepted, rejected or otherwise).

1. Theme localization (l10n). This has been completed in #2299 and will be available in version 1.2.0.
2. Support multi-language search. This has been **completed** and is already [available](https://www.mkdocs.org/user-guide/configuration/#search).
3. Add support for pages in multiple languages. There are currently two proposals for how to do this: #774 which proposes a system in which pages are manually translated, and #617 which proposes a system in which pages are auto-translated using po files or the like. Some third party plugins exist to address these issues. It is likely that any native solution would come from a third party plugin if/when one establishes itself as well developed and is a clear favorite among users. Whether that means the plugin would be moved in-house or it would become the "recommended" solution and remain a third-party plugin will depend upon the circumstances at the time that such a decision is made.

At the time of this edit, no one on the MkDocs core team has a need for multi-language support. Therefore, this is not a high priority for us, The best way for you to move any of these issues forward is to volunteer your time and energy working on the issues. We will do our best to review PRs and shepherd discussions.

_&lt;/edit&gt;_

It would be great if mkdocs supports i18n(Internationalization), just like [gitbook](https://github.com/GitbookIO/gitbook). We can `mkdir` en and others. Then add a button in navbar to change the language which changes the docs_dir in fact.

I have found some solutions for jinjia2 i18n. Is it possible to implement with that?

**Comment by m-baumgartner** [comment]
What about simplified version of this where `mkdocs` just provide a selector to the reader to switch between the detected languages (by detecting language-code-named `.md` files for instance) and the user handle the translation of the markdown documents himself and provide them to `mkdocs`? 
https://crowdin.com/ for instance provide a localization plateform which supports markdown formated files.
That would a first step probably easier and quicker to deliver?

**Comment by d0ugal** [comment]
We don't want to add a half baked solution in the interim. This needs to be solved at some point in the future with standard tooling.

**Comment by Unknwon** [comment]
+1

**Comment by Surgo** [comment]
+1

**Comment by ecarreras** [comment]
We have developed a [markdown-i18n plugin](https://github.com/gisce/markdown-i18n) and we use one mkdocs.yml config for each language.

I know is not the best solution (and the plugin is a Work In Progress code) but for now it works for our needs

**Comment by tanner0101** [comment]
Any updates on this or ideas on how the community could help? We currently use MKDocs for https://docs.vapor.codes/3.0/ and we would love to localize them.

**Comment by waylan** [comment]
There are (at least) three parts to supporting multiple languages, which *should* be worked on in this order:

1. Theme localization (l10n). In other words the templates need to be updated to support translations of the next/previous` navigation, page footer, etc.
2. Support multi-language search. This was **completed** out-of-order and is [already available][1].
3. Add support for pages in multiple languages. There are currently two proposals for how to do this: #774 which proposes a system in which pages are manually translated, and #617 which proposes a system in which pages are auto-translated using po files or the like. Either solution will likely need to be some wrapper which runs a separate MkDocs build for each language similar to how [jimporter/mike][2] runs a separate build for each version. 

I expect step 1 to be the easiest to do. Someone just needs to do the work. Step 2 was done out-of-order because I happened to be working on search anyway and it was easier to incorporate multi-language support from the get-go. Step 3 is likely to be the most difficult and doesn't really make sense if we don't have step 1.  Step 3 can also easily be a third party tool.

In any event, as I understand it, no one on the MkDocs team has any need for multi-language support. Therefore, this is not a high priority for us, The best way to move this forward is for a volunteer to start working on step 1 and submit a PR which I'm happy to review.

[1]: https://www.mkdocs.org/user-guide/configuration/#search
[2]:  https://github.com/jimporter/mike

**Comment by SheKnowsItsAMultipass** [comment]
> There are (at least) three parts to supporting multiple languages, which _should_ be worked on in this order:
> 
> 1. Theme localization (l10n). In other words the templates need to be updated to support translations of the next/previous` navigation, page footer, etc.
> 2. Support multi-language search. This was **completed** out-of-order and is [already available](https://www.mkdocs.org/user-guide/configuration/#search).
> 3. Add support for pages in multiple languages. There are currently two proposals for how to do this: #774 which proposes a system in which pages are manually translated, and #617 which proposes a system in which pages are auto-translated using po files or the like. Either solution will likely need to be some wrapper which runs a separate MkDocs build for each language similar to how [jimporter/mike](https://github.com/jimporter/mike) runs a separate build for each version.
> 
> I expect step 1 to be the easiest to do. Someone just needs to do the work. Step 2 was done out-of-order because I happened to be working on search anyway and it was easier to incorporate multi-language support from the get-go. Step 3 is likely to be the most difficult and doesn't really make sense if we don't have step 1. Step 3 can also easily be a third party tool.
> 
> In any event, as I understand it, no one on the MkDocs team has any need for multi-language support. Therefore, this is not a high priority for us, The best way to move this forward is for a volunteer to start working on step 1 and submit a PR which I'm happy to review.

We would have a need for this and I was thinking simpler (I think) with no need for step 3.  We could have a folder for our English .md files (docs) and a folder for our French .md files (docs-fr).  This seems the most flexible for us as its the most intuitive way to create the doc for each language with screenshots, etc...   

We could then call 'mkdocs build' and 'mkdocs build fr' to basically generate two satitc sites.

There might be a need to do something similar with the .yml file.

So to recap, the templates need to be localized and a new command line parameter would be needed.

**Comment by waylan** [comment]
@DaveLHX you seem to have the general idea. At the very least, with step 1 done a user always could simply have multiple MkDocs instances (one for each language). The new command line stuff could always some later.

**Comment by DaveLHX** [comment]
> @DaveLHX you seem to have the general idea. At the very least, with step 1 done a user always could simply have multiple MkDocs instances (one for each language). The new command line stuff could always some later.

How far along is step 1?  it seems to waiting on some actions for a few months.  I dont need this feature right now (but it would be a good selling point for my managers at work) and after your comment I realize I could make the two mkdocs folders and add a manually translated version the template and then generate the two sites.  Not sure if this would catch everything but feels like I could manage.

**Comment by waylan** [comment]
Step 1 has not been started. We need a volunteer to work on it.

**Comment by DaveLHX** [comment]
> Step 1 has not been started. We need a volunteer to work on it.

I thought it was this one https://github.com/mkdocs/mkdocs/pull/1778
I may be confusing things.

**Comment by waylan** [comment]
I stand corrected. Sorry, I forgot about that. Yes, the work in #1778 is the beginning of step 1.

**Comment by TheMulti0** [comment]
Is this ever going to be solved?

**Comment by waylan** [comment]
@TheMulti0 as noted above, the work was started in #1778. However, the person who started it seems to have abandoned it before it was completed. There is a itemized list there of what needs to be completed to get this all working. When all of the items lists are completed, then it can be merged and will be available in the next release.

So to answer your question, this will be "solved" when someone does the work. Personally, I have no need for the feature and I have very limited time to work on any new features for MkDocs. At most, I can only provide the time required to review someone else's work on this issue. In fact, we all are volunteers who work on this in our spare time. If you would like to see this resolved, then I would suggest contributing some time to resolve the outstanding items in #1778 yourself.

**Comment by TheMulti0** [comment]
I understand, thanks for replying!

**Comment by sim6** [comment]
I wrote this plugin https://gitlab.com/mkdocs-i18n/mkdocs-i18n/ to meet the requirements of https://isardvdi.com project (https://isard.gitlab.io/isardvdi-docs/). I hope it is useful for you. It could be adapted to other requirements, feel free to open problems at https://gitlab.com/mkdocs-i18n/mkdocs-i18n/-/issues.

**Comment by waylan** [comment]
I am excited to announce that theme localization has just been added in #2299 and will be available in version 1.2.0 of MkDocs!

We will leave this issue open until we decide definitively how we will move forward with translation of page content. There are currently two proposals for how to do this: #774  and #617. A third option may be to leave this for third-party plugins to handle.

**Comment by 9bow** [comment]
Great, thanks!
I look forward to the release of 1.2.0!

**Comment by micwoj92** [comment]
@waylan this issue says i18n but the mentioned #2299 pr is l10n, why not support additional 8 languages? or I understand something wrong...

**Comment by trel** [comment]
These are different things.

https://lokalise.com/blog/internationalization-vs-localization/

**Comment by ultrabug** [comment]
@micwoj92 we indeed debated a little on i18n vs l10n naming.

While we initially thought that l10n was maybe appropriate, the i18n term imposed itself to stay in line with the underlying Jinja2 i18n extension we base upon.

It is worth mentioning that i18n has also been used by mkdocs plugins to add support for page translations such as https://github.com/ultrabug/mkdocs-static-i18n

**Comment by ultrabug** [comment]
i18n is now supported by mkdocs themes and there are [plugins for content localization available](https://github.com/mkdocs/mkdocs/wiki/MkDocs-Plugins#internationalization--localization) that should be considered (mkdocs-static-i18n is most [mature at the time of writing this comment](https://github.com/mkdocs/mkdocs/issues/774#issuecomment-1207802122) - disclaimer: I'm the author of the mkdocs-static-i18n plugin).
